File: gateway-service/app/sqs-graphql-client.go
package main

import (
	"fmt"
	"io"
	"os"
	"strings"

	"net/http"

	"github.com/Khan/genqlient/graphql"
	"github.com/TwiN/go-color"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/sqs"
)

type SQSTransport struct {
	Headers                map[string]string
	SQSService             *sqs.SQS
	MessageDeduplicationId string
	MessageGroupId         string
	QueueUrl               string
}

const (
	BillingFifoQueueUrl = "https://sqs.us-east-1.amazonaws.com/887279901853/graphql-service-billing-queue.fifo"
	UsageLogQueueUrl    = "https://sqs.us-east-1.amazonaws.com/887279901853/graphql-service-usage-log-queue"
)

func (self *SQSTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	for k, v := range self.Headers {
		req.Header.Add(k, v)
	}
	bodyBytes, _ := io.ReadAll(req.Body)
	body := string(bodyBytes)
	// fmt.Println("body", string(body))
	// fmt.Println(color.Green, "Sending request url", req.URL, req.Host, color.Reset)
	// fmt.Println(color.Green, "Sending request body", string(body), color.Reset)
	// fmt.Println(color.Green, "Sending request header", req.Header, color.Reset)
	var messageGroupIdPtr *string
	if self.MessageGroupId != "" {
		messageGroupIdPtr = &self.MessageGroupId
	}
	var messageDeduplicationIdPtr *string
	if self.MessageDeduplicationId != "" {
		messageDeduplicationIdPtr = &self.MessageDeduplicationId
	}
	_, err := self.SQSService.SendMessage(&sqs.SendMessageInput{
		MessageBody:            &body,
		MessageGroupId:         messageGroupIdPtr,
		MessageDeduplicationId: messageDeduplicationIdPtr,
		QueueUrl:               &self.QueueUrl,
	})
	if err != nil {
		fmt.Println(color.Red, "Error sending SQS message", err, color.Reset)
		return nil, err
	}
	// fmt.Println(color.Yellow, "Received response", resp, color.Reset)
	// fmt.Println(color.Yellow, "Received response err", err, color.Reset)
	return &http.Response{
		StatusCode: 200,
		Body:       io.NopCloser(strings.NewReader("{}")),
	}, nil
}

type SQSGraphQLClientConfig struct {
	MessageDeduplicationId string
	MessageGroupId         string
	QueueUrl               string
}

// Returns a GraphQL client that uses SQS as a transport layer.
//
// MessageDeduplicationId: Only for fifo queues. Sets the MessageDeduplicationId
// for the SQS message.
//
// MessageGroupId: Only for fifo queues. Sets the MessageGroupId for the SQS
// message delay_seconds: selects the SQS queue to use based on the
// delay_seconds. For the billing queue, this must be "main".
//
// Effectively, MessageDeduplicationId deduplicates messages with the same ID.
// Messages with the same MessageGroupId are processed in FIFO order.
func getSQSGraphQLClient(config SQSGraphQLClientConfig) *graphql.Client {
	if os.Getenv("LOCAL_GRAPHQL") == "1" {
		fmt.Println(color.Red, "Using local GraphQL in place of SQS", color.Reset)
		return getGraphQLClient() // Use normal GraphQL client (local in this case)
	}

	// Doc: https://github.com/sha1sum/aws_signing_client
	sess := session.Must(session.NewSession(&aws.Config{
		Region: aws.String("us-east-1"),
	}))
	if config.QueueUrl == BillingFifoQueueUrl {
		config.MessageGroupId = "main"
	}
	baseClient := http.Client{
		Transport: &SQSTransport{
			Headers: map[string]string{
				"X-Service-Name": "gateway",
			},
			SQSService:             sqs.New(sess),
			MessageDeduplicationId: config.MessageDeduplicationId,
			MessageGroupId:         config.MessageGroupId,
			QueueUrl:               config.QueueUrl,
		},
	}
	client := graphql.NewClient("", &baseClient)
	return &client
}

File: cli/tests/sqs_graphql_client.py
from enum import Enum
from json import dumps as json_dumps
from typing import Optional

import boto3
from fastchargeapi_cli.config import aws_account_id
from gql import Client
from gql.transport.requests import RequestsHTTPTransport
from requests import Response


class PredefinedSQSQueue(Enum):
    billing_fifo_queue = f"https://sqs.us-east-1.amazonaws.com/{aws_account_id}/graphql-service-billing-queue.fifo"
    usage_log_queue = f"https://sqs.us-east-1.amazonaws.com/{aws_account_id}/graphql-service-usage-log-queue"


class SQSHttpTransport(RequestsHTTPTransport):
    def __init__(
        self,
        url: str,
        queue: PredefinedSQSQueue,
        MessageDeduplicationId: Optional[str] = None,
        MessageGroupId: Optional[str] = None,
        **kwargs,
    ):
        super().__init__(url=url, **kwargs)
        self.MessageDeduplicationId = MessageDeduplicationId
        self.MessageGroupId = MessageGroupId
        self.queue = queue

    def connect(self):
        self.session = SQSRequestsSession(
            MessageDeduplicationId=self.MessageDeduplicationId,
            MessageGroupId=self.MessageGroupId,
            queue=self.queue,
        )


class SQSRequestsSession:
    def __init__(
        self,
        queue: PredefinedSQSQueue,
        MessageDeduplicationId: Optional[str] = None,
        MessageGroupId: Optional[str] = None,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.MessageDeduplicationId = MessageDeduplicationId
        self.MessageGroupId = MessageGroupId
        self.queue = queue
        self.client = boto3.client("sqs")

    def mount(self):
        pass

    def close(self):
        pass

    def request(
        self,
        method: str,
        url: str,
        json: object = None,
        **kwargs,
    ):
        response = self.client.send_message(
            QueueUrl=self.queue.value,
            MessageBody=json_dumps(json),
            MessageDeduplicationId=self.MessageDeduplicationId,
            MessageGroupId=self.MessageGroupId,
        )
        resp = Response()
        resp.status_code = 200
        resp._content = b'{"data": {}}'
        return resp


def get_sqs_graphql_client(
    queue: PredefinedSQSQueue,
    dedup_id: Optional[str] = None,
    group_id: str = "main",
) -> Client:
    """Returns a GraphQL client that uses SQS as a transport layer.
    dedup_id: Only for fifo queues. Sets the MessageDeduplicationId for the SQS
    message.

    group_id: Only for fifo queues. Sets the MessageGroupId for the SQS message.
    If the billing is PredefinedSQSQueue.billing_fifo_queue, you need to use
    "main" as the group_id.

    Effectively, MessageDeduplicationId deduplicates messages with the same ID.
    Messages with the same MessageGroupId are processed in FIFO order.
    """
    if queue == PredefinedSQSQueue.billing_fifo_queue:
        assert group_id and group_id.startswith(
            "user_"
        ), "When using the billing queue, use the user's pk as the group_id."

    return Client(
        transport=SQSHttpTransport(
            url="",
            queue=queue,
            MessageDeduplicationId=dedup_id,
            MessageGroupId=group_id,
        ),
        fetch_schema_from_transport=False,
    )
